File: coreutils-rs/util/src/meta.rs
/*
 * SPDX-License-Identifier: GPL-3.0-or-later
 *
 * This file is part of coreutils-rs from scratch.
 * Copyright (c) 2025 Horstaufmental
 *
 * coreutils-rs from scratch is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * coreutils-rs from scratch is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 */
pub struct ProgramMeta {
    pub name: &'static str,
    pub project: &'static str,
    pub version: &'static str,
    pub authors: &'static str,
}

pub fn print_version(meta: &ProgramMeta) {
    println!("{} ({}) {}", meta.name, meta.project, meta.version);
    println! {"Copyright (C) 2025 {}", meta.authors};
    println!(
        "License GPLv3+: GNU GPL version 3 or later \
         <https://gnu.org/licenses/gpl.html>.\n\
         This is free software: you are free to change and redistribute it.\n\
         There is NO WARRANTY, to the extent permitted by law.\n"
    );
    println!("Written by {}", meta.authors);
}

File: shared/meta.c
#include "meta.h"

#include <stdio.h>
#include <string.h>

#define BOLD_ON \033[1m
#define BOLD_OFF \033[0m

#define STRINGIFY(x) #x
#define TOSTRING(x) STRINGIFY(x)

void print_help(const char *usage, const char *desc,
                const struct help_entry *help_entries, const char *footer) {
  puts(usage);
  puts(desc);

  // find longest option string
  int maxlen = 0;
  for (int i = 0; help_entries[i].opt; i++) {
    int len = (int)strlen(help_entries[i].opt);
    if (len > maxlen)
      maxlen = len;
  }

  // print each option aligned
  for (int i = 0; help_entries[i].opt; i++) {
    if (help_entries[i].placement == Inline)
      printf("  " TOSTRING(BOLD_ON) "%-*s" TOSTRING(BOLD_OFF) "  %s\n", maxlen, help_entries[i].opt, help_entries[i].desc);
    else
      printf("  " TOSTRING(BOLD_ON) "%-*s" TOSTRING(BOLD_OFF) "\n         %s\n", maxlen, help_entries[i].opt, help_entries[i].desc);
  }

  if (footer != NULL) {
    putchar('\n');
    puts(footer);
  }
}

void print_version(const char *prog_name, const char *proj_name,
                   const char *ver, const char *authors) {
  printf("%s (%s) %s\n", prog_name, proj_name, ver);
  printf("Copyright (C) 2025 %s\n", authors);
  puts("License GPLv3+: GNU GPL version 3 or later "
       "<https://gnu.org/licenses/gpl.html>.\n"
       "This is free software: you are free to change and redistribute it.\n"
       "There is NO WARRANTY, to the extent permitted by law.\n");
  printf("Written by %s\n", authors);
}
